File: baekjoon_all/04000+/boj_4485.cpp
// Solve 2023-11-24
// Update 2024-03-26

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;
using pii = pair<int, int>;

const int INF = 1e9;

struct Point{
    int x, y, cost;

    Point(int x, int y, int cost) : x(x), y(y), cost(cost) {}

    bool operator<(const Point &rhs) const {
        if (cost != rhs.cost) cost < rhs.cost;
        if (x != rhs.x) return x < rhs.x;
        return y < rhs.y;
    }

    bool operator>(const Point &rhs) const {
        return !(*this < rhs);
    }
};

int main() {
    FASTIO;

    int dx[4] = { 0, 0, 1, -1 };
    int dy[4] = { 1, -1, 0, 0 };
    int ti = 0;

    while (true) {
        int n;
        cin >> n;

        if (n == 0) break;

        vector<vector<int>> board(n, vector<int>(n, 0));

        for (vector<int> &row : board) {
            for (int &col : row) {
                cin >> col;
            }
        }

        vector<vector<int>> min_dist(n, vector<int>(n, INF));
        priority_queue<Point, vector<Point>, greater<Point> > pq_mintop;
        min_dist[0][0] = board[0][0];
        pq_mintop.emplace(0, 0, board[0][0]);

        while (!pq_mintop.empty()) {
            Point cur = pq_mintop.top();
            pq_mintop.pop();

            if (min_dist[cur.x][cur.y] < cur.cost) continue;
            if (cur.x == n - 1 && cur.y == n - 1) break;

            for (int d = 0; d < 4; d++) {
                int nx = cur.x + dx[d];
                int ny = cur.y + dy[d];

                if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;

                int ncost = cur.cost + board[nx][ny];

                if (min_dist[nx][ny] > ncost) {
                    min_dist[nx][ny] = ncost;
                    pq_mintop.emplace(nx, ny, ncost);
                }
            }
        }

        cout << "Problem " << ++ti << ": " << min_dist[n - 1][n - 1] << '\n';
    }

    return 0;
}

File: baekjoon_all/04000+/boj_4485.java
// Solve 2024-03-26

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class boj_4485 {

    static class Point {
        int x, y, cost;

        Point(int x, int y, int cost) {
            this.x = x;
            this.y = y;
            this.cost = cost;
        }
    }

    static int[] dx = { 0, 0, 1, -1 };
    static int[] dy = { 1, -1, 0, 0 };
    static int[][] board = new int[125][125];
    static int[][] minDist = new int[125][125];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int ti = 0;

        while (true) {
            int n = Integer.parseInt(br.readLine());

            if (n == 0) {
                break;
            }

            for (int x = 0; x < n; x++) {
                st = new StringTokenizer(br.readLine(), " ");

                for (int y = 0; y < n; y++) {
                    board[x][y] = Integer.parseInt(st.nextToken());
                    minDist[x][y] = 1_000_000_000;
                }
            }

            PriorityQueue<Point> pqMinTop = new PriorityQueue<Point>(new Comparator<Point>() {

                @Override
                public int compare(Point o1, Point o2) {
                    return o1.cost < o2.cost ? -1 : o1.cost > o2.cost ? 1 : 0;
                }

            });

            minDist[0][0] = board[0][0];
            pqMinTop.offer(new Point(0, 0, board[0][0]));

            while (!pqMinTop.isEmpty()) {
                Point cur = pqMinTop.poll();

                if (cur.cost > minDist[cur.x][cur.y]) {
                    continue;
                }

                if (cur.x == n - 1 && cur.y == n - 1) {
                    break;
                }

                for (int d = 0; d < 4; d++) {
                    int nx = cur.x + dx[d];
                    int ny = cur.y + dy[d];

                    if (nx < 0 || nx >= n || ny < 0 || ny >= n) {
                        continue;
                    }

                    int nDist = cur.cost + board[nx][ny];

                    if (nDist < minDist[nx][ny]) {
                        minDist[nx][ny] = nDist;
                        pqMinTop.offer(new Point(nx, ny, nDist));
                    }
                }
            }

            sb.append("Problem ").append(++ti).append(": ").append(minDist[n - 1][n - 1]).append("\n");
        }

        System.out.print(sb);
        br.close();
    }

}
